### PR TITLE
Feature/json report

### PR DESCRIPTION
**Pull request by sshivaji**
Slimmer PR for just JSON report for ESBMC. 

To use:

esbmc file--unwind 10 --branch-coverage --generate-json-report

### CODE REVIEW

**Review by mikhailramalho — CHANGES_REQUESTED** [review]
Please go through Rafael's comment from the last PR. Stuff like global variables, leftover comments, and unnecessary includes need to be fixed.

**Comment by sshivaji** [comment]
Thx @mikhailramalho, I have made changes

File: src/goto-symex/json.cpp [hunk 1]
```diff
@@ -0,0 +1,386 @@
...
+            case type2t::fixedbv_id: return "fixedbv";
+            case type2t::unsignedbv_id: return "unsigned";
+            case type2t::signedbv_id: return "signed";
+            default: return fmt::format("unknown({})", static_cast<int>(id));
+        }
+    }
+
+    std::set<std::string> find_included_headers(const std::string& file_path, std::set<std::string>& processed_files) {
+        if (processed_files.find(file_path) != processed_files.end() || 
+            starts_with(std::filesystem::path(file_path).string(), "/usr/")) {
```
Comment (rafaelsamenezes) [inline src/goto-symex/json.cpp]: I think you are the same check a few times now. Consider adding a function and documenting why the /usr is the only prefix to check

**Review by rafaelsamenezes — CHANGES_REQUESTED** [review]
Just one comment. Otherwise LGTM.

**Comment by sshivaji** [comment]
Now updated for codacy and @rafaelsamenezes's comments, thanks!

**Review by lucasccordeiro — APPROVED** [review]

File: src/goto-symex/json.cpp [hunk 2]
```diff
@@ -0,0 +1,405 @@
...
+            }
+            test_entry["source_files"] = source_data;
+        }
+
+        all_tests.push_back(test_entry);
+
+        // Write updated JSON
+        std::ofstream json_out("report.json");
+        if(!json_out.is_open()) {
+            log_status("Error: Could not open report.json for writing");
```
Comment (ChenfengWei0) [inline src/goto-symex/json.cpp]: `log_error()` might be more appropriate for this case

File: src/goto-symex/json.cpp [hunk 3]
```diff
@@ -0,0 +1,405 @@
...
+        json all_tests = []() {
+            // TODO: Allow diff filename for report
+            std::ifstream input("report.json");
+            if(input.is_open()) {
+                try {
+                    json existing;
+                    input >> existing;
+                    return existing;
+                } catch(json::parse_error& e) {
+                    log_status("Error parsing existing report.json: {}", e.what());
```
Comment (ChenfengWei0) [inline src/goto-symex/json.cpp]: Same as above

File: src/goto-symex/json.cpp [hunk 4]
```diff
@@ -0,0 +1,405 @@
...
+                            }
+                        }
+                        return struct_data;
+                    }
+                    else {
+                        return from_expr(ns, "", expr);
+                    }
+                }
+                catch(const std::runtime_error& e) {
+                    log_status("Serialization error: {}", e.what());
```
Comment (ChenfengWei0) [inline src/goto-symex/json.cpp]: same as above

File: src/goto-symex/json.cpp [hunk 5]
```diff
@@ -0,0 +1,405 @@
...
+                            };
+                        }
+                        else {
+                            step_data["type"] = "other";
+                        }
+
+                        test_entry["steps"].push_back(step_data);
+                    }
+                } catch(std::exception& e) {
+                    log_status("Error processing step: {}", e.what());
```
Comment (ChenfengWei0) [inline src/goto-symex/json.cpp]: Same as above

File: src/goto-symex/json.cpp [hunk 6]
```diff
@@ -0,0 +1,405 @@
...
+
+        // Handle source files for first entry
+        if(all_tests.empty()) {
+            json source_data;
+            for(const auto& file : all_referenced_files) {
+                if(!is_system_path(file)) {
+                    try {
+                        source_data[file] = read_file_lines(file);
+                    } catch(std::exception& e) {
+                        log_status("Error reading file {}: {}", file, e.what());
```
Comment (ChenfengWei0) [inline src/goto-symex/json.cpp:378]: Same as above

**Review by ChenfengWei0 — APPROVED** [review]
Overall LGTM, just some minor correction

File: src/goto-symex/json.cpp (hunk 6 above)
Comment (sshivaji) [reply src/goto-symex/json.cpp:378]: Thanks, will def change to this

Comment (sshivaji) [reply src/goto-symex/json.cpp:378]: Done!

**Comment by lucasccordeiro** [comment]
Thanks for submitting this PR, @sshivaji.

**Pull request merged**